### PR TITLE
tooling: Add requirements.in and rationalize deps

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: tooling: Add requirements.in and rationalize deps
Additional Description:

This adds a `requirements.in` file to list actual deps to create the `requirements.txt` file from, which will make it *much* easier to manage pip deps

It also removes several requirements targets that are not necessary

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by htuch** [comment]
@phlax please try split PRs like this in the future into trivial (i.e. the requirements.in) which can be rubber stamped and the unrelated deps rationalization.

File: tools/testing/requirements.in [hunk 1]
```diff
@@ -0,0 +1,5 @@
+coverage
```
Comment (htuch) [inline tools/testing/requirements.in:1]: We're not picking specific versions any more?

File: tools/testing/requirements.txt [hunk 2]
```diff
@@ -64,61 +62,50 @@ coverage==5.5 \
...
-    #   -r tools/testing/requirements.txt
-    #   pytest
-pluggy==1.0.0 \
-    --hash=sha256:74134bbf457f031a36d68416e1509f34bd5ccc019f0bcc952c7b909d06b37bd3 \
-    --hash=sha256:4224373bacce55f955a878bf9cfa763c1e360858e330072059e10bad68531159
-    # via
-    #   -r tools/testing/requirements.txt
-    #   pytest
+    # via pytest
+pluggy==0.13.1 \
```
Comment (htuch) [inline tools/testing/requirements.txt:75]: It looks like we downgraded the `pluggy` version here..

**Review by htuch — COMMENTED** [review]
@moderation for comment.

File: tools/testing/requirements.txt (hunk 2 above)
Comment (phlax) [reply tools/testing/requirements.txt:75]: i think that is required - there was a conflict with pytest, and a hurried fix this week - ill check again tho

File: tools/testing/requirements.in (hunk 1 above)
Comment (phlax) [reply tools/testing/requirements.in:1]: not in the `requirements.in` file unless we have a specific reason to - the `requirements.txt` file is generated from these and that does lock/pin/pick specific versions

that said we can set versions here if we need to and `pip-compile` should respect them (not sure about dependabot)

**Comment by phlax** [comment]
> ... and the unrelated deps rationalization.

its not really unrelated, i added the pins so that i could do the rationalization - what i was reluctant to do was add loads of pins, and requirements file updates (which would have been just as large a pr) when rationalizing just got rid of the files altogether

i have a few prs in waiting to further tidy up and rationalize tooling, they all intersect, which is why i pushed this up front, to get rid of anything that would otherwise be repeatedly changed

File: tools/testing/requirements.txt (hunk 2 above)
Comment (phlax) [reply tools/testing/requirements.txt:75]: yep, re-running pip-compile gives the ~same result, so i think that is the current deps situation

Comment (phlax) [reply tools/testing/requirements.txt]: my bad - i wasnt upgrading when running pip-compile - ive reverted the pluggy downgrade

File: tools/deprecate_features/deprecate_features.py [hunk 3]
```diff
@@ -4,7 +4,6 @@
 import re
 import subprocess
 import fileinput
-from six.moves import input
```
Comment (phlax) [inline tools/deprecate_features/deprecate_features.py:7]: if we wan to keep `six` here we could depend on it from elsewhere (ie `base_pip3`)

i would prefer not to. Python >= 3.8 is an assumption baked into a lot of the tooling (even tho much will work regardless)

the pytooling repo tests against 3.8 and 3.9 currently

File: tools/testing/requirements.in (hunk 1 above)
Comment (phlax) [reply tools/testing/requirements.in:1]: some further investigation of using `pip-compile` 

one annoying issue is that it uses sets and doesnt order the hashes, which means running it is non-deterministic (ie the hashes keep reordering) - which is a pita 8/. I had thought that this problem was due to running different versions of pip-compile, but now it seems like the issue is more as described.

the other thing which doesnt work so well with pip-compile is conditional deps

if you set a conditional dep on eg `sys_platform = "win32"` or somesuch pip-compile respects it in the sense that if you are not on windows it doesnt get included in the requirements.txt file

this is pretty rubbish for mananging repo requirements if you expect different install/run environments

Comment (phlax) [reply tools/testing/requirements.in:1]: this is already an issue as windows seems to be expecting `pyreadline` in some of my prs

we can just add it as dep, currently that would make linux pull it in unnecessarily, but that also could be resolved with (the new) pip_install_incremental in bazel

the problem with this approach occurs where there is a dep for some particular env that you really dont want to be pulled in, but im not sure that is an issue for envoy. In that case the best thing would be to separate the deps in bazel i think

**Comment by phlax** [comment]
#18026 provides a better solution for this as it rationlizes the requirements down to a single file by using the new pip_install_incrememental

im therefore closing this in favour of that pr

**Pull request closed without merging**